File: frontend/src/services/bookingService.js
// src/services/bookingService.js
import api from './api';

const BOOKING_BASE = '/bookings';

export const bookingService = {
    // Create a booking
    createBooking: (data) => api.post(BOOKING_BASE, data).then(res => res.data),

    // Get user's own bookings
    getMyBookings: (params) => api.get(`${BOOKING_BASE}/my`, { params }).then(res => res.data),

    // Get booking by ID
    getBookingById: (id) => api.get(`${BOOKING_BASE}/${id}`).then(res => res.data),

    // Cancel a booking
    cancelBooking: (id) => api.delete(`${BOOKING_BASE}/${id}/cancel`).then(res => res.data),

    // Get available time slots for a resource on a date
    getAvailableTimeSlots: (resourceId, date) =>
        api.get(`${BOOKING_BASE}/available-slots`, { params: { resourceId, date } }).then(res => res.data),

    // Check if a time slot conflicts
    checkConflict: (resourceId, date, startTime, endTime) =>
        api.get(`${BOOKING_BASE}/check-conflict`, { params: { resourceId, date, startTime, endTime } }).then(res => res.data),

    // Admin: Get all bookings
    getAllBookings: (params) => api.get(BOOKING_BASE, { params }).then(res => res.data),

    // Admin: Update booking status
    updateBookingStatus: (id, data) => api.patch(`${BOOKING_BASE}/${id}/status`, data).then(res => res.data)
};

export default bookingService;

File: campus-operations-hub/src/main/java/com/smartcampus/operations/service/BookingService.java
// src/main/java/com/smartcampus/operations/service/BookingService.java
package com.smartcampus.operations.service;

import com.smartcampus.operations.dto.*;
import com.smartcampus.operations.entity.BookingStatus;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.Pageable;

import java.time.LocalDate;
import java.time.LocalTime;
import java.util.UUID;

public interface BookingService {

    BookingResponse createBooking(BookingCreateRequest request, String userEmail);

    BookingResponse updateBookingStatus(UUID bookingId, BookingStatusUpdateRequest request, String adminEmail);

    BookingResponse cancelBooking(UUID bookingId, String userEmail);

    // Updated method signature with 3 parameters
    Page<BookingResponse> getUserBookings(String userEmail, String status, String bookingType, Pageable pageable);

    Page<BookingResponse> getAllBookings(UUID resourceId, String status, LocalDate bookingDate, Pageable pageable);

    AvailableTimeSlotsResponse getAvailableTimeSlots(UUID resourceId, LocalDate date);

    boolean checkConflict(UUID resourceId, LocalDate date, LocalTime startTime, LocalTime endTime);

    BookingResponse getBookingById(UUID bookingId, String userEmail);

    BookingResponse updateBooking(UUID bookingId, BookingUpdateRequest request, String userEmail);

    // Dashboard count methods
    long getTotalBookingsCount();
    long getPendingBookingsCount();
    long getApprovedBookingsCount();
    long getRejectedBookingsCount();
    long getCancelledBookingsCount();
    long getMaintenanceBookingsCount();

    // User-specific counts
    long getUserBookingsCount(String userEmail);
    long getUserPendingBookingsCount(String userEmail);
    long getUserApprovedBookingsCount(String userEmail);
}